pkg/prometheus/promcfg.go

```diff
@@ -2142,6 +2181,37 @@ func (cg *ConfigGenerator) generateK8SSDConfig(
 	}
 }
 
+func (cg *ConfigGenerator) generateRoleSelectorConfig(k8sSDConfig yaml.MapSlice, roles []string, selector metav1.LabelSelector) yaml.MapSlice {
+	selectors := []yaml.MapSlice{}
+	for _, role := range roles {
+		yml := yaml.MapSlice{
+			{
+				Key:   "role",
+				Value: role,
+			},
+		}
+
+		labelSelector := labels.SelectorFromValidatedSet(labels.Set(selector.MatchLabels))
+
+		if len(selector.MatchExpressions) > 0 {
+			for _, exp := range selector.MatchExpressions {
+				// Needs to lower because the selection expects lowercase.
+				// exp.Operator is CamelCase.
+				requirement, err := labels.NewRequirement(exp.Key, selection.Operator(strings.ToLower(string(exp.Operator))), exp.Values)
+				if err != nil {
+					cg.logger.Error("failed to create label requirement", "err", err)
+					continue
+				}
+				labelSelector = labelSelector.Add(*requirement)
+			}
+		}
+		yml = append(yml, yaml.MapItem{Key: "label", Value: labelSelector.String()})
+		selectors = append(selectors, yml)
+	}
+
+	return cg.AppendMapItem(k8sSDConfig, "selectors", selectors)
```

Comment:
should we check if the "selectors" is already present in k8sSDConfig, just to be safe?